qiskit/transpiler/passes/adjacent_barriers.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+"""
+A pass that merges any adjacent barriers into one
+"""
+
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.extensions.standard.barrier import Barrier
+
+
+class MergeAdjacentBarriers(TransformationPass):
+    """Returns a circuit with any adjacent barriers merged together"""
+
+    def run(self, dag):
+
+        # sorted to so that they are in the order they were added to the DAG
+        # so ancestors/descendants makes sense
+        barriers = sorted(dag.named_nodes('barrier'))
```

Comment:
To sort the barriers on the graph, we'd need a topological sort, since we're concerned with their ordering in the DAG, which is not necessarily the same as the order in which they were added to the DAG.

Reply:
Yes I agree, will update this.